transport/bdp_estimator.go

```diff
@@ -0,0 +1,47 @@
+package transport
+
+import (
+	"sync"
+)
+
+type bdpEstimator struct {
+	p                *ping
+	maxWindow        uint32
+	mu               sync.Mutex
+	estimate         uint32
+	accumulator      uint32 // Number of bytes received.
+	isSent           bool
+	send             func(*ping)    // Callback to send ping.
+	updateConnWindow func(n uint32) // Callback to update window size.
+}
+
+func (b *bdpEstimator) add(n uint32) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	if !b.isSent {
+		b.isSent = true
+		b.accumulator = 0
+		b.send(b.p)
+	}
+	b.accumulator += n
+}
+
+func (b *bdpEstimator) calculate(d [8]byte) {
+	if b.p.data != d {
+		return
+	}
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	b.isSent = false
+	estimate := 2 * b.accumulator
+	if estimate > b.estimate {
```

Comment:
We found it necessary to make sure the bandwidth also increased, in order to avoid buffer bloat in the network. Buffer bloat was most severe in testing setups, but lives in the wild quite a bit as well.

https://github.com/grpc/grpc-java/blob/v1.2.0/netty/src/main/java/io/grpc/netty/AbstractNettyHandler.java#L180